File: mobile/ios/Sources/Browser2030BCore/SecurityPolicy.swift
import Foundation

/// Capabilities a site or feature can request. Everything is denied by default
/// (Security/Privacy spec §3), matching the Rust engine's capability model.
public enum Capability: String, Sendable, CaseIterable {
    case network
    case geolocation
    case camera
    case microphone
    case notifications
    case aiLocal
    case aiRemote
    case telemetry
}

/// Policy layers in increasing precedence.
public enum PolicyLayer: Int, Sendable, Comparable {
    case `default` = 0
    case user = 1
    case localAdmin = 2
    case enterprise = 3

    public static func < (lhs: PolicyLayer, rhs: PolicyLayer) -> Bool {
        lhs.rawValue < rhs.rawValue
    }
}

/// A single capability grant.
public struct Grant: Sendable, Equatable {
    public let capability: Capability
    public let layer: PolicyLayer
    public let allowed: Bool
    /// Seconds since the reference date after which the grant lapses; nil = session.
    public let expiresAt: TimeInterval?

    public init(
        capability: Capability,
        layer: PolicyLayer,
        allowed: Bool,
        expiresAt: TimeInterval? = nil
    ) {
        self.capability = capability
        self.layer = layer
        self.allowed = allowed
        self.expiresAt = expiresAt
    }
}

/// Default-deny security policy. Not thread-safe by design — wrap in an actor at
/// the call site if shared across tasks (kept simple for unit testing).
public struct SecurityPolicy: Sendable {
    private var grants: [Capability: Grant] = [:]

    public init() {}

    /// Record a grant if it comes from an equal-or-higher layer than any existing.
    public mutating func set(_ grant: Grant) {
        if let existing = grants[grant.capability], grant.layer < existing.layer {
            return
        }
        grants[grant.capability] = grant
    }

    /// A capability is allowed only by an unexpired allow grant (default-deny).
    public func isAllowed(_ capability: Capability, now: TimeInterval = Date().timeIntervalSinceReferenceDate) -> Bool {
        guard let g = grants[capability] else { return false }
        if let expiry = g.expiresAt, now >= expiry { return false }
        return g.allowed
    }

    public mutating func revoke(_ capability: Capability) {
        grants[capability] = nil
    }
}

File: mobile/android/src/main/kotlin/ai/b2030b/android/SecurityPolicy.kt
package ai.b2030b.android

/**
 * Default-deny security policy for the Android edition (Security/Privacy spec §3).
 *
 * Mirrors the Rust engine's capability model: every permission, network egress,
 * AI feature, and telemetry sink starts denied. The host app must explicitly
 * grant a capability (subject to admin policy) before it is honored.
 *
 * Kept free of Android framework types so it is unit-testable on the JVM.
 */
enum class Capability {
    NETWORK,
    GEOLOCATION,
    CAMERA,
    MICROPHONE,
    NOTIFICATIONS,
    AI_LOCAL,
    AI_REMOTE,
    TELEMETRY,
}

/** Where a decision came from, in increasing precedence. */
enum class PolicyLayer { DEFAULT, USER, LOCAL_ADMIN, ENTERPRISE }

data class Grant(
    val capability: Capability,
    val layer: PolicyLayer,
    val allowed: Boolean,
    /** Epoch millis after which the grant lapses; null = session. */
    val expiresAt: Long? = null,
)

class SecurityPolicy {
    private val grants = mutableMapOf<Capability, Grant>()

    /** Record a grant, but only if it comes from an equal-or-higher layer. */
    fun set(grant: Grant) {
        val existing = grants[grant.capability]
        if (existing == null || grant.layer.ordinal >= existing.layer.ordinal) {
            grants[grant.capability] = grant
        }
    }

    /** Default-deny query. A capability is allowed only by an unexpired allow. */
    fun isAllowed(capability: Capability, now: Long = System.currentTimeMillis()): Boolean {
        val g = grants[capability] ?: return false
        if (g.expiresAt != null && now >= g.expiresAt) return false
        return g.allowed
    }

    fun revoke(capability: Capability) {
        grants.remove(capability)
    }
}
